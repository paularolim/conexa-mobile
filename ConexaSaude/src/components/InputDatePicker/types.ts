export interface InputDatePickerProps {
  testID?: string;
  label: string;
  value: Date;
  displayedValue?: string;
  isModalOpen: boolean;
  openModal: () => void;
  onConfirm: (date: Date) => void;
  onCancel: () => void;
}
