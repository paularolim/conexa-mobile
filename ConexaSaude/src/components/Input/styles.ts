import { TextInput, View } from 'react-native';
import styled from 'styled-components/native';

import { Text } from '@components/Text';

export const Container = styled(View)`
  background-color: ${({ theme }) => theme.colors.white};
  width: 100%;
  justify-content: center;
  align-items: flex-start;
`;

export const Label = styled(Text)`
  color: ${({ theme }) => theme.colors.gray};
  font-size: 18px;
`;

export const FieldContainer = styled(View)`
  border: 1px solid ${({ theme }) => theme.colors.gray};
  border-radius: 4px;
  width: 100%;
  flex-direction: row;
  align-items: center;
  min-height: 48px;
`;

export const Field = styled(TextInput)`
  color: ${({ theme }) => theme.colors.black};
  flex-grow: 1;
  padding: 8px 16px;
  flex: 1;
`;

export const IconContainer = styled(View)`
  padding-right: 16px;
`;
