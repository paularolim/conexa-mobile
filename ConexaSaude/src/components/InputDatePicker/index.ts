export * from './InputDatePicker';
export * from './types';
