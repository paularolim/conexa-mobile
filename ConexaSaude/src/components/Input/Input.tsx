import { useState } from 'react';
import { TouchableOpacity } from 'react-native';
import { useTheme } from 'styled-components/native';

import { Icon } from '@components/Icon';

import { Container, Field, FieldContainer, Label } from './styles';
import { InputProps } from './types';

export function Input({ label, hasPasswordIcon, secureTextEntry, ...rest }: InputProps) {
  const [hidePassword, setHidePassword] = useState(secureTextEntry);

  const { colors } = useTheme();

  const togglePasswordVisibility = () => {
    setHidePassword((prev) => !prev);
  };

  return (
    <Container>
      <Label>{label}</Label>
      <FieldContainer>
        <Field testID="field" secureTextEntry={hidePassword} {...rest} />
        {hasPasswordIcon && (
          <TouchableOpacity onPress={togglePasswordVisibility} testID="password-visibility-toggle">
            <Icon
              name={hidePassword ? 'eye-on' : 'eye-off'}
              width={25}
              height={25}
              fill={colors.gray}
            />
          </TouchableOpacity>
        )}
      </FieldContainer>
    </Container>
  );
}
