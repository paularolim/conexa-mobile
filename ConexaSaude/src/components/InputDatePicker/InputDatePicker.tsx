import { Pressable, View } from 'react-native';
import DatePicker from 'react-native-date-picker';

import { Input } from '@components/Input/Input';

import { InputDatePickerProps } from './types';

export function InputDatePicker({
  testID = 'input-date-picker',
  label,
  value,
  displayedValue,
  isModalOpen,
  openModal,
  onCancel,
  onConfirm,
}: InputDatePickerProps) {
  return (
    <>
      <Pressable onPress={openModal}>
        <View pointerEvents="none">
          <Input label={label} value={displayedValue || `${value}`} editable={false} />
        </View>
      </Pressable>
      <DatePicker
        testID={`${testID}-modal`}
        modal
        open={isModalOpen}
        date={value}
        onConfirm={onConfirm}
        onCancel={onCancel}
      />
    </>
  );
}
