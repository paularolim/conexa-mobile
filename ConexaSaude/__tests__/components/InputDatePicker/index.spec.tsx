import { render, fireEvent } from '@testing-library/react-native';
import DatePicker from 'react-native-date-picker';
import renderer from 'react-test-renderer';

import { Input } from '@components/Input';
import { InputDatePicker } from '@components/InputDatePicker';

import { Wrapper } from '../../Wrapper';

jest.mock('react-native-date-picker', () => jest.fn());
(DatePicker as jest.Mock).mockImplementation((props) => <Input label={props.label} {...props} />);

describe('InputDatePicker component', () => {
  it('should render component correctly', () => {
    const tree = renderer
      .create(
        <Wrapper>
          <InputDatePicker
            label="Test Input"
            isModalOpen={false}
            value={new Date()}
            displayedValue="2023-10-10"
            openModal={jest.fn()}
            onConfirm={jest.fn()}
            onCancel={jest.fn()}
          />
        </Wrapper>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render with label and displayedValue', () => {
    const mockProps = {
      label: 'Date',
      value: new Date(),
      displayedValue: '2023-10-10',
      isModalOpen: false,
      openModal: jest.fn(),
      onCancel: jest.fn(),
      onConfirm: jest.fn(),
    };

    const { getByText, getByDisplayValue } = render(
      <Wrapper>
        <InputDatePicker {...mockProps} />
      </Wrapper>,
    );

    expect(getByText(mockProps.label)).toBeTruthy();
    expect(getByDisplayValue(mockProps.displayedValue)).toBeTruthy();
  });

  it('should render value if displayedValue is not provided', () => {
    const mockProps = {
      label: 'Date',
      value: new Date(),
      isModalOpen: false,
      openModal: jest.fn(),
      onCancel: jest.fn(),
      onConfirm: jest.fn(),
    };

    const { getByDisplayValue } = render(
      <Wrapper>
        <InputDatePicker {...mockProps} />
      </Wrapper>,
    );

    expect(getByDisplayValue(`${mockProps.value}`)).toBeTruthy();
  });

  it('should call openModal when pressed', () => {
    const mockProps = {
      label: 'Date',
      value: new Date(),
      displayedValue: '2023-10-10',
      isModalOpen: false,
      openModal: jest.fn(),
      onCancel: jest.fn(),
      onConfirm: jest.fn(),
    };

    const { getByTestId } = render(
      <Wrapper>
        <InputDatePicker {...mockProps} />
      </Wrapper>,
    );

    fireEvent.press(getByTestId('field'));

    expect(mockProps.openModal).toHaveBeenCalledTimes(1);
  });

  it('should render the DatePicker modal when isModalOpen is true', () => {
    const mockProps = {
      label: 'Date',
      value: new Date(),
      displayedValue: '2023-10-10',
      isModalOpen: true,
      openModal: jest.fn(),
      onCancel: jest.fn(),
      onConfirm: jest.fn(),
    };

    const { getByTestId } = render(
      <Wrapper>
        <InputDatePicker {...mockProps} />
      </Wrapper>,
    );

    expect(getByTestId('input-date-picker-modal')).toBeTruthy();
  });
});
