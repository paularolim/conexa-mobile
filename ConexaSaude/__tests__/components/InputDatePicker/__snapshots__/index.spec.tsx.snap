// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputDatePicker component should render component correctly 1`] = `
[
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <View
      pointerEvents="none"
    >
      <View
        style={
          [
            {
              "alignItems": "flex-start",
              "backgroundColor": "#FFFFFF",
              "justifyContent": "center",
              "width": "100%",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#000000",
              },
              {
                "color": "#707683",
                "fontSize": 18,
              },
            ]
          }
        >
          Test Input
        </Text>
        <View
          style={
            [
              {
                "alignItems": "center",
                "borderColor": "#707683",
                "borderRadius": 4,
                "borderStyle": "solid",
                "borderWidth": 1,
                "flexDirection": "row",
                "minHeight": 48,
                "width": "100%",
              },
            ]
          }
        >
          <TextInput
            editable={false}
            style={
              [
                {
                  "color": "#000000",
                  "flexBasis": 0,
                  "flexGrow": 1,
                  "flexShrink": 1,
                  "paddingBottom": 8,
                  "paddingLeft": 16,
                  "paddingRight": 16,
                  "paddingTop": 8,
                },
              ]
            }
            testID="field"
            value="2023-10-10"
          />
        </View>
      </View>
    </View>
  </View>,
  <View
    style={
      [
        {
          "alignItems": "flex-start",
          "backgroundColor": "#FFFFFF",
          "justifyContent": "center",
          "width": "100%",
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "color": "#000000",
          },
          {
            "color": "#707683",
            "fontSize": 18,
          },
        ]
      }
    />
    <View
      style={
        [
          {
            "alignItems": "center",
            "borderColor": "#707683",
            "borderRadius": 4,
            "borderStyle": "solid",
            "borderWidth": 1,
            "flexDirection": "row",
            "minHeight": 48,
            "width": "100%",
          },
        ]
      }
    >
      <TextInput
        date={2023-10-04T12:21:45.926Z}
        modal={true}
        onCancel={[MockFunction]}
        onConfirm={[MockFunction]}
        open={false}
        style={
          [
            {
              "color": "#000000",
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "paddingBottom": 8,
              "paddingLeft": 16,
              "paddingRight": 16,
              "paddingTop": 8,
            },
          ]
        }
        testID="input-date-picker-modal"
      />
    </View>
  </View>,
]
`;
